lib/modules.nix

```diff
@@ -87,31 +90,70 @@ rec {
-        else
-          let file = toString x; key = toString x; in
-          unifyModuleSyntax file key (applyIfFunction key (import x) args));
-    in
-      builtins.genericClosure {
-        startSet = toClosureList unknownModule "" modules;
-        operator = m: toClosureList m._file m.key m.imports;
-      };
+      # Collects all modules recursively into the form
+      #
+      #   {
+      #     disabled = [ <list of disabled modules> ];
+      #     keys.key1 = {
+      #       module = <module for key1>;
+      #       keys.key1-2 = {
+      #         module = <module for key1-2>;
+      #         keys = ...;
+      #       };
+      #     };
+      #   }
+      #
+      # - `disabled` contains the list of all disabled modules
+      # - `keys` is an attribute containing the module at each key along with all
+      #   modules keys that were imported by that module
+      collectStructuredModules =
+        let
+          collectResults = list: {
+            disabled = concatLists (catAttrs "disabled" list);
+            keys = listToAttrs list;
+          };
+        in parentFile: parentKey: modules: args: collectResults (imap1 (n: x:
+          let
+            module = fullyUnifyModuleSyntax args parentFile "${parentKey}:anon-${toString n}" x;
+            collectedImports = collectStructuredModules module._file module.key module.imports args;
+          in {
+            name = module.key;
+            disabled = module.disabledModules ++ collectedImports.disabled;
+            value = { module = module; keys = collectedImports.keys; };
+          }) modules);
+
+      # filterModules :: String -> { disabled, keys } -> [ Module ]
+      #
+      # Filters a structure as emitted by collectStructuredModules by removing all disabled
+      # modules recursively. It returns the final list of unique-by-key modules
+      filterModules = modulesPath: { disabled, keys }:
+        let
+          moduleKey = m: if isString m then toString modulesPath + "/" + m else toString m;
+          disabledKeys = map moduleKey disabled;
+          # Takes a recursive `keys` set as described in collectStructuredModules and flattens
+          # it out to a direct `key -> module` mapping while also filtering disabled
+          # keys
+          flattenKeys = keys:
+            let included = filterAttrs (key: value: ! elem key disabledKeys) keys;
+            in mapAttrs (key: value: value.module) included
+              // foldl' (a: b: a // flattenKeys b.keys) {} (attrValues included);
+              # By using the //, we are deduplicating modules with the same key
+        in attrValues (flattenKeys keys);
+
+    in modulesPath: modules: args:
+      filterModules modulesPath (collectStructuredModules unknownModule "" modules args);
+
+  # Like unifyModuleSyntax, but also imports paths and calls functions if necessary
+  fullyUnifyModuleSyntax = args: fallbackFile: fallbackKey: m:
```

Comment:
This seems to have a higher-level goal than unifying syntax. What about `loadModule`?